Subject: Support macOS 27 on the official yabai source

Base: https://github.com/asmvik/yabai/commit/dd845723416f5fe92af49fad5ebab00369e07edd
Source: https://github.com/AhsanFazal/yabai/commit/ad0a12d63f639534a296a1d065b0d04979f1b4db
Initial support: https://github.com/AhsanFazal/yabai/commit/49c2d035aa0a7d99f6ebf7f7d327972d6a707bb4
Scripting addition version: https://github.com/AhsanFazal/yabai/commit/6f4990479f94d0706a327e9571cd6b3adb32be10
Upstream tracking: https://github.com/asmvik/yabai/issues/2802

Keep the upstream macOS 26.6 changes and omit the fork's local installer.

diff --git a/src/workspace.h b/src/workspace.h
--- a/src/workspace.h
+++ b/src/workspace.h
@@ -2,12 +2,13 @@
 #define WORKSPACE_H
 
 #define SUPPORTED_MACOS_VERSION_LIST    \
-    SUPPORT_MACOS_VERSION(tahoe,    26) \
-    SUPPORT_MACOS_VERSION(sequoia,  15) \
-    SUPPORT_MACOS_VERSION(sonoma,   14) \
-    SUPPORT_MACOS_VERSION(ventura,  13) \
-    SUPPORT_MACOS_VERSION(monterey, 12) \
-    SUPPORT_MACOS_VERSION(bigsur,   11)
+    SUPPORT_MACOS_VERSION(goldengate,   27) \
+    SUPPORT_MACOS_VERSION(tahoe,        26) \
+    SUPPORT_MACOS_VERSION(sequoia,      15) \
+    SUPPORT_MACOS_VERSION(sonoma,       14) \
+    SUPPORT_MACOS_VERSION(ventura,      13) \
+    SUPPORT_MACOS_VERSION(monterey,     12) \
+    SUPPORT_MACOS_VERSION(bigsur,       11)
 
 #define SUPPORT_MACOS_VERSION(name, major_version) \
 static bool _workspace_is_macos_version_##name; \
diff --git a/src/event_loop.c b/src/event_loop.c
--- a/src/event_loop.c
+++ b/src/event_loop.c
@@ -18,7 +18,7 @@
     int window_count = 0;
     uint32_t window_list[1024] = {0};
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         // NOTE(asmvik): Subscribe to all windows because of window_destroyed (and ordered) notifications
         table_for (struct window *window, g_window_manager.window, {
             window_list[window_count++] = window->id;
@@ -242,7 +242,7 @@
         view_clear_flag(view, VIEW_IS_DIRTY);
     }
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         update_window_notifications();
     }
 }
@@ -336,7 +336,7 @@
         view_clear_flag(view, VIEW_IS_DIRTY);
     }
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         update_window_notifications();
     }
 
@@ -595,7 +595,7 @@
         event_signal_push(SIGNAL_WINDOW_CREATED, window);
     }
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         update_window_notifications();
     }
 }
@@ -628,7 +628,7 @@
     window_unobserve(window);
     window_destroy(window);
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         update_window_notifications();
     }
 }
@@ -1550,7 +1550,7 @@
         workspace_is_macos_ventura() ||
         workspace_is_macos_sonoma() ||
         workspace_is_macos_sequoia() ||
-        workspace_is_macos_tahoe()) {
+        (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         mission_control_unobserve();
         mission_control_observe();
     }
diff --git a/src/view.c b/src/view.c
--- a/src/view.c
+++ b/src/view.c
@@ -41,7 +41,7 @@
         SLSReenableUpdate(g_connection);
         SLSOrderWindow(g_connection, node->feedback_window.id, 1, node->window_order[0]);
         table_add(&g_window_manager.insert_feedback, &node->window_order[0], node);
-        if (!workspace_is_macos_sequoia() && !workspace_is_macos_tahoe()) {
+        if (!workspace_is_macos_sequoia() && !(workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
             update_window_notifications();
         }
     }
@@ -107,7 +107,7 @@
     if (node->feedback_window.id) {
         table_remove(&g_window_manager.insert_feedback, &node->window_order[0]);
 
-        if (!workspace_is_macos_sequoia() && !workspace_is_macos_tahoe()) {
+        if (!workspace_is_macos_sequoia() && !(workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
             update_window_notifications();
         }
 
diff --git a/src/yabai.c b/src/yabai.c
--- a/src/yabai.c
+++ b/src/yabai.c
@@ -312,13 +312,13 @@
         workspace_is_macos_ventura() ||
         workspace_is_macos_sonoma() ||
         workspace_is_macos_sequoia() ||
-        workspace_is_macos_tahoe()) {
+        (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         mission_control_observe();
 
         if (workspace_is_macos_ventura() ||
             workspace_is_macos_sonoma() ||
             workspace_is_macos_sequoia() ||
-            workspace_is_macos_tahoe()) {
+            (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
             SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1327, NULL);
             SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1328, NULL);
         }
@@ -329,7 +329,7 @@
     SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 808, NULL);
     SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1202, NULL);
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 804, NULL);
     }
 
@@ -337,7 +337,7 @@
     space_manager_begin(&g_space_manager);
     window_manager_begin(&g_space_manager, &g_window_manager);
 
-    if (workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         update_window_notifications();
     }
 
diff --git a/src/window.c b/src/window.c
--- a/src/window.c
+++ b/src/window.c
@@ -900,7 +900,7 @@
 {
     int level = 0;
 
-    if (workspace_is_macos_ventura() || workspace_is_macos_sonoma() || workspace_is_macos_sequoia() || workspace_is_macos_tahoe()) {
+    if (workspace_is_macos_ventura() || workspace_is_macos_sonoma() || workspace_is_macos_sequoia() || (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         CFArrayRef window_ref = cfarray_of_cfnumbers(&wid, sizeof(uint32_t), 1, kCFNumberSInt32Type);
 
         CFTypeRef query = SLSWindowQueryWindows(g_connection, window_ref, 1);
@@ -945,7 +945,7 @@
     msg.header.msgh_bits = 0x1513;
     msg.header.msgh_remote_port = CGSGetConnectionPortById(cid);
     msg.header.msgh_local_port = mig_get_special_reply_port();
-    msg.header.msgh_id = workspace_is_macos_tahoe() ? 0x76E3 : 0x73C3;
+    msg.header.msgh_id = (workspace_is_macos_tahoe() || workspace_is_macos_goldengate()) ? 0x76E3 : 0x73C3;
     mach_msg(&msg.header, MACH_SEND_MSG|MACH_RCV_MSG, 0x24, 0x30, msg.header.msgh_local_port, 0, 0);
 
     return msg.sub_level;
diff --git a/src/window_manager.c b/src/window_manager.c
--- a/src/window_manager.c
+++ b/src/window_manager.c
@@ -2264,7 +2264,7 @@
         workspace_is_macos_ventura() ||
         workspace_is_macos_sonoma() ||
         workspace_is_macos_sequoia() ||
-        workspace_is_macos_tahoe()) {
+        (workspace_is_macos_tahoe() || workspace_is_macos_goldengate())) {
         while (!space_is_user(space_manager_active_space())) {
 
             //
diff --git a/src/osax/arm64_payload.m b/src/osax/arm64_payload.m
--- a/src/osax/arm64_payload.m
+++ b/src/osax/arm64_payload.m
@@ -5,7 +5,7 @@
     __asm__("mov x0, %0\n""mov x1, %1\n""mov x2, %2\n""mov x20, %3\n" : :"r"(v0), "r"(v1), "r"(v2), "r"(v3) :"x0", "x1", "x2", "x20"); ((void (*)())(func))();
 
 uint64_t get_dock_spaces_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         return 0x30000;
     } else if (os_version.majorVersion == 15) {
         return os_version.minorVersion >= 4 ? 0x1f0000 : 0x200000;
@@ -21,7 +21,9 @@
 }
 
 uint64_t get_dppm_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return 0x40000;
+    } else if (os_version.majorVersion == 26) {
         return 0x70000;
     } else if (os_version.majorVersion == 15) {
         return 0x250000;
@@ -37,7 +39,9 @@
 }
 
 uint64_t get_fix_animation_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return 0x210000;
+    } else if (os_version.majorVersion == 26) {
         return 0x250000;
     } else if (os_version.majorVersion == 15) {
         return 0x250000;
@@ -53,7 +57,9 @@
 }
 
 uint64_t get_add_space_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return 0x210000;
+    } else if (os_version.majorVersion == 26) {
         return 0x250000;
     } else if (os_version.majorVersion == 15) {
         return 0x250000;
@@ -69,7 +75,9 @@
 }
 
 uint64_t get_remove_space_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return 0x170000;
+    } else if (os_version.majorVersion == 26) {
         return 0x1e0000;
     } else if (os_version.majorVersion == 15) {
         return 0x1c0000;
@@ -85,7 +93,9 @@
 }
 
 uint64_t get_move_space_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return 0x170000;
+    } else if (os_version.majorVersion == 26) {
         return 0x1c0000;
     } else if (os_version.majorVersion == 15) {
         return 0x1c0000;
@@ -101,7 +111,7 @@
 }
 
 uint64_t get_set_front_window_offset(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         return 0x10000;
     } else if (os_version.majorVersion == 15) {
         return 0x35000;
@@ -117,7 +127,7 @@
 }
 
 const char *get_dock_spaces_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         // Pulling out of doBindingCommand:display (search decompiled text in ghidra) function.
         return "?8 ?? ?? ?? 08 ?? ?? 91 00 01 40 F9 E2 03 13 AA ?? ?? ?? 94 ?? ?? ?? ?? 08";
     } else if (os_version.majorVersion == 15) {
@@ -137,7 +147,7 @@
 }
 
 const char *get_dppm_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         //Pulling from function 'DPRemoteConnection::_handleEvent:'
         return "?? ?? 00 ?? 08 ?? ?? 91 00 01 40 F9 E2 03 16 AA E3 03 19 AA ?? ?? ?? 94";
     } else if (os_version.majorVersion == 15) {
@@ -157,7 +167,7 @@
 }
 
 const char *get_fix_animation_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         return "00 10 6A 1E A8 ?? ?? D1 ?? 01 ?? F8";
     } else if (os_version.majorVersion == 15) {
         return "00 10 6A 1E A8 ?? ?? D1 ?? 01 ?? F8";
@@ -173,7 +183,9 @@
 }
 
 const char *get_add_space_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return "7F 23 03 D5 E1 03 1E AA ?? ?? ?? 97 FE 03 01 AA ?? ?? ?? A9 FD 43 01 91 F3 03 14 AA F5 03 00 AA 96 8E 43 F8";
+    } else if (os_version.majorVersion == 26) {
         if (os_version.minorVersion >= 4) {
             return "7F 23 03 D5 E1 03 1E AA ?? ?? ?? 97 FE 03 01 AA FD 7B 05 A9 FD 43 01 91 F3 03 14 AA F5 03";
         }
@@ -192,7 +204,7 @@
 }
 
 const char *get_remove_space_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         return "7F 23 03 D5 FF ?? ?? D1 FC ?? ?? A9 FA ?? ?? A9 F8 ?? ?? A9 F6 ?? ?? A9 F4 ?? ?? A9 FD ?? ?? A9 FD ?? ?? 91 ?? 03 03 AA F5 03 02 AA F4 03 01 AA";
     } else if (os_version.majorVersion == 15) {
         return "7F 23 03 D5 FF 83 ?? D1 FC 6F ?? A9 FA 67 ?? A9 F8 5F ?? A9 F6 57 ?? A9 F4 4F ?? A9 FD 7B ?? A9 FD 43 ?? 91 ?? 03 03 AA ?? 03 02 AA ?? 03 01 AA ?? 03 00 AA ?? ?? ?? AA";
@@ -208,7 +220,7 @@
 }
 
 const char *get_move_space_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 26 || os_version.majorVersion == 27) {
         return "7F 23 03 D5 E3 03 1E AA ?? ?? ?? 97 FE 03 03 AA FD 7B ?? A9 FD ?? ?? 91 F6 03 14 AA";
     } else if (os_version.majorVersion == 15) {
         return "7F 23 03 D5 E3 03 1E AA ?? ?? FF 97 FE 03 03 AA FD 7B 06 A9 FD 83 01 91 F6 03 14 AA F4 03 02 AA FB 03 01 AA FA 03 00 AA ?? 13 00 ?? E8 ?? ?? F9 19 68 68 F8 E0 03 19 AA E1 03 16 AA";
@@ -228,7 +240,9 @@
 }
 
 const char *get_set_front_window_pattern(NSOperatingSystemVersion os_version) {
-    if (os_version.majorVersion == 26) {
+    if (os_version.majorVersion == 27) {
+        return "?? ?? ?? 34 7F 23 03 D5 FF C3 01 D1 ?? ?? ?? A9 ?? ?? ?? A9 ?? ?? ?? A9 FD 83 01 91 F3 03 01 AA F5 03 00 AA";
+    } else if (os_version.majorVersion == 26) {
         return "21 ?? ?? 34 7F 23 03 D5 FF ?? 01 D1 F6 ?? 04 A9 F4 ?? 05 A9 FD ?? 06 A9 FD ?? 01 91";
     } else if (os_version.majorVersion == 15) {
         return "7F 23 03 D5 FF ?? 02 D1 F6 57 ?? A9 F4 4F ?? A9 FD 7B ?? A9 FD ?? 02 91 ?? ?? 00 ?? 08 ?? ?? F9";
diff --git a/src/osax/payload.m b/src/osax/payload.m
--- a/src/osax/payload.m
+++ b/src/osax/payload.m
@@ -238,6 +238,11 @@
         NSLog(@"[yabai-sa] Detected Tahoe Preview... flagging 'macOSSequoia=true.'");
         macOSSequoia = true;
         return true; // Tahoe preview
+    } else if (os_version.majorVersion == 27) {
+
+        NSLog(@"[yabai-sa] Detected Golden Gate Developer Beta (26A5353q, 26A5368g, 26A5378j, 26A5378n, 26A5388g, 26A5406e, 26A5416b)... flagging 'macOSSequoia=true.'");
+        macOSSequoia = true;
+        return true; // Golden Gate Developer Beta (26A5353q, 26A5368g, 26A5378j, 26A5378n, 26A5388g, 26A5406e, 26A5416b)
     }
 
     NSLog(@"[yabai-sa] spaces functionality is only supported on macOS Big Sur 11.0.0+, Monterey 12.0.0+, Ventura 13.0.0+, Sonoma 14.0.0+, and Sequoia 15.0");
@@ -256,6 +261,11 @@
         NSLog(@"[yabai-sa] Detected Tahoe Preview... flagging 'macOSSequoia=true.'");
         macOSSequoia = true;
         return true; // Tahoe preview
+    } else if (os_version.majorVersion == 27) {
+
+        NSLog(@"[yabai-sa] Detected Golden Gate Developer Beta (26A5353q, 26A5368g, 26A5378j, 26A5378n, 26A5388g, 26A5406e, 26A5416b)... flagging 'macOSSequoia=true.'");
+        macOSSequoia = true;
+        return true; // Golden Gate Developer Beta (26A5353q, 26A5368g, 26A5378j, 26A5378n, 26A5388g, 26A5406e, 26A5416b)
     }
 
     NSLog(@"[yabai-sa] spaces functionality is only supported on macOS Monterey 12.0.0+, and Ventura 13.0.0+, Sonoma 14.0.0+, and Sequoia 15.0");
diff --git a/src/osax/common.h b/src/osax/common.h
--- a/src/osax/common.h
+++ b/src/osax/common.h
@@ -4,7 +4,7 @@
 #define SA_SOCKET_PATH_FMT "/tmp/yabai-sa_%s.socket"
 #define SA_SOCKET_BUFF_LEN 0x1000
 
-#define OSAX_VERSION                "2.1.30"
+#define OSAX_VERSION                "2.1.31"
 
 #define OSAX_ATTRIB_DOCK_SPACES     0x01
 #define OSAX_ATTRIB_DPPM            0x02
